plugins/deye_plugin_ha_discovery.py

```diff
@@ -258,24 +280,43 @@ def publish_sensor_information(self, topic: str, observation: Observation):
         # <discovery_prefix>/<component>/[<node_id>/]<object_id>/config
         discovery_topic = f"{discovery_prefix}/sensor/{node_id}/{object_id}/config"
 
-        discover_config = {
-            "name": observation.sensor.name,
-            "unique_id": self._get_unique_id(observation.sensor.name),
-            "force_update": True,
-            "device_class": device_class,
-            "state_class": state_class,
-            "unit_of_measurement": self._adapt_unit(observation.sensor.unit),
-            "availability_topic": f"{self._config.mqtt.topic_prefix}/status",
-            "state_topic": topic,
-            "device": {
-                "identifiers": [node_id],
-                "name": self._device_name,
-                "manufacturer": self.inverter_manufacturer,
-                "model": f"{self.inverter_model} SN:{self._logger_serial}",
-                "serial_number": str(self._logger_serial),
-                "sw_version": f"deye-inverter-mqtt with {self.get_id()}",
-            },
-        }
+        if device_class == "enum":
+            discover_config = {
+                "name": observation.sensor.name,
+                "unique_id": self._get_unique_id(observation.sensor.name),
+                "force_update": True,
+                "device_class": device_class,
+                "options": enum_options,
+                "availability_topic": f"{self._config.mqtt.topic_prefix}/status",
+                "state_topic": topic,
+                "device": {
+                    "identifiers": [node_id],
+                    "name": self._device_name,
+                    "manufacturer": self.inverter_manufacturer,
+                    "model": f"{self.inverter_model} SN:{self._logger_serial}",
+                    "serial_number": str(self._logger_serial),
+                    "sw_version": f"deye-inverter-mqtt with {self.get_id()}",
+                },
+            }
+        else:
+            discover_config = {
+                "name": observation.sensor.name,
+                "unique_id": self._get_unique_id(observation.sensor.name),
+                "force_update": True,
+                "device_class": device_class,
+                "state_class": state_class,
+                "unit_of_measurement": self._adapt_unit(observation.sensor.unit),
+                "availability_topic": f"{self._config.mqtt.topic_prefix}/status",
+                "state_topic": topic,
+                "device": {
+                    "identifiers": [node_id],
+                    "name": self._device_name,
+                    "manufacturer": self.inverter_manufacturer,
+                    "model": f"{self.inverter_model} SN:{self._logger_serial}",
+                    "serial_number": str(self._logger_serial),
+                    "sw_version": f"deye-inverter-mqtt with {self.get_id()}",
+                },
+            }
```

Comment:
What do you think about rollback to the previous code and update the `discover_config` structure in a second step?

Untested example:
```suggestion
        discover_config = {
            "name": observation.sensor.name,
            "unique_id": self._get_unique_id(observation.sensor.name),
            "force_update": True,
            "device_class": device_class,
            "state_class": state_class,
            "unit_of_measurement": self._adapt_unit(observation.sensor.unit),
            "availability_topic": f"{self._config.mqtt.topic_prefix}/status",
            "state_topic": topic,
            "device": {
                "identifiers": [node_id],
                "name": self._device_name,
                "manufacturer": self.inverter_manufacturer,
                "model": f"{self.inverter_model} SN:{self._logger_serial}",
                "serial_number": str(self._logger_serial),
                "sw_version": f"deye-inverter-mqtt with {self.get_id()}",
            },
        }
        if device_class == "enum":
            del discover_config["unit_of_measurement"]
            discover_config["options"] = enum_options
```

Reply:
I totally agree, that was quick&dirty. Improved in ce0026a (tested).